### PR TITLE
[#9536] Instructor Feedback Sessions Page E2E Test 

### LINKED ISSUE #9536: Add E2E testing to teammates pages in v7
Migrate end to end tests for all the pages below from v6 to v7

- [x] Student Home Page
- [x] Student Profile Page
- [x] Student Course Details Page
- [x] Admin Home Page
- [x] Admin Accounts Page
- [x] Admin Search Page
- [x] Admin Sessions Page
- [x] Instructor Home Page
- [x] Instructor Course Details Page
- [x] Instructor Courses Page
- [x] Instructor Course Edit Page
- [x] Instructor Course Enroll Page
- [x] Instructor Course Student Details Page
- [x] Instructor Course Student Edit Page
- [x] Instructor Search Page (ticked by default; see division of cases)
  - [x] Students table
  - [ ] Responses + response comments table
- [x] Instructor Student List Page
- [x] Instructor Student Records Page (ticked by default; see division of cases)
  - [x] Student details
  - [ ] Feedback responses given/received by student
- [x] Instructor Session Edit Page
- [x] Instructor Sessions Page
- [x] Instructor Session Result Page
- [x] Session Submission Page
- [x] Session Result Page
- [x] Course Join Page

### PR DESCRIPTION
**Pull request by jtankw3**
Part of #9536

**Outline of Solution**

- Add methods to `BaseE2ETestCase` to verify downloaded files exist
- Add methods to `BaseE2ETestCase` to compare emails sent
- Add `InstructorFeedbackSessionsPage` to represent instructor sessions page
- Add relevant backdoor methods to compare feedback session data in datastore
- Add `InstructorFeedbackSessionsPageE2ETest` to test:
    - verify loaded data
    - verify response rate
    - add new session
    - add new copied session
    - copy session
    - publish results
    - send reminder email
    - resend results link
    - unpublish results
    - download results
    - soft delete session
    - restore session
    - permanently delete session
    - restore all session
    - delete all session

### CODE REVIEW

File: src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java [hunk 1]
```diff
@@ -146,4 +157,45 @@ protected AdminHomePage loginAdmin() {
...
+        try {
+            email.getUserAuthenticated();
+            int retryLimit = 5;
+            boolean actual = email.isEmailWithSubjectPresent(subject);
+            while (!actual && retryLimit > 0) {
+                retryLimit--;
+                ThreadHelper.waitFor(1000);
+                actual = email.isEmailWithSubjectPresent(subject);
+            }
+            email.markAllUnreadEmailAsRead();
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java:194]: Is this necessary?

File: src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java [hunk 2]
```diff
@@ -146,4 +157,45 @@ protected AdminHomePage loginAdmin() {
...
+        String filePath = getTestDownloadsFolder() + fileName;
+        FileHelper.deleteFile(filePath);
+    }
+
+    /**
+     * Verify file with fileName is in the downloads folder.
+     */
+    protected void verifyDownloadedFile(String expectedFileName) {
+        Path path = Paths.get(getTestDownloadsFolder() + expectedFileName);
+        assertTrue(Files.exists(path));
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java:173]: Is it feasible to also check the content (or a subset of) of the file?

File: src/e2e/java/teammates/e2e/pageobjects/Browser.java [hunk 3]
```diff
@@ -58,6 +62,7 @@
     public Browser() {
         this.driver = createWebDriver();
         this.driver.manage().window().maximize();
+        this.driver.manage().timeouts().pageLoadTimeout(TestProperties.TEST_TIMEOUT * 2, TimeUnit.SECONDS);
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/Browser.java:65]: 👍

File: src/e2e/java/teammates/e2e/pageobjects/InstructorFeedbackSessionsPage.java [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+        fillTextBox(copyFsModal.findElement(By.id("copy-session-name")), newSessionName);
+        selectCourseToCopyToInModal(copyFsModal, copyToCourse.getId());
+
+        click(browser.driver.findElement(By.id("btn-confirm-copy-course")));
+    }
+
+    public void moveToRecycleBin(FeedbackSessionAttributes sessionToDelete) {
+        int rowId = getFeedbackSessionRowId(sessionToDelete.getCourseId(), sessionToDelete.getFeedbackSessionName());
+        clickAndConfirm(browser.driver.findElement(By.id("btn-soft-delete-" + rowId)));
+        ThreadHelper.waitFor(1000);
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/InstructorFeedbackSessionsPage.java:207]: What are all these waits for?

File: src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java (hunk 1 above)
Comment (jtankw3) [reply src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java:203]: It is not really necessary but is there because the checks in `EmailAccount` only looks at unread mails and so the mail is marked as read after the test to reduce the searching.

File: src/e2e/java/teammates/e2e/pageobjects/InstructorFeedbackSessionsPage.java (hunk 4 above)
Comment (jtankw3) [reply src/e2e/java/teammates/e2e/pageobjects/InstructorFeedbackSessionsPage.java]: It is actually for the animation time when the tables are created. I have changed to using the `waitUntilAnimationFinish` method that someone else wrote.

File: src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java [hunk 5]
```diff
@@ -146,4 +158,50 @@ protected AdminHomePage loginAdmin() {
...
+    }
+
+    /**
+     * Compares file with fileName in the downloads folder with expected file in results folder.
+     */
+    protected void verifyDownloadedFileContent(String expectedFileName, String actualFileName) {
+        try {
+            String expectedFilePath = getTestResultsFolder() + expectedFileName;
+            String actualFilePath = getTestDownloadsFolder() + actualFileName;
+            assertEquals(FileHelper.readFile(expectedFilePath), FileHelper.readFile(actualFilePath));
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java:176]: This works, but we don't really need to do a full check (otherwise we need to also maintain these test files). Just doing a quick sanity check (file name and maybe a few representative lines) is good enough.

File: src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java [hunk 6]
```diff
@@ -146,4 +153,49 @@ protected AdminHomePage loginAdmin() {
...
+        }
+        EmailAccount emailAccount = new EmailAccount(email);
+        try {
+            emailAccount.getUserAuthenticated();
+            int retryLimit = 5;
+            boolean actual = emailAccount.isEmailWithSubjectPresent(subject);
+            while (!actual && retryLimit > 0) {
+                retryLimit--;
+                ThreadHelper.waitFor(1000);
+                actual = emailAccount.isEmailWithSubjectPresent(subject);
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java:192]: I have seen a lot of this kind of operation, and it's not just on this PR. Is it possible to build some kind of abstraction for them? Or perhaps use the `RetryManager` class? But don't need to do on this PR.

File: src/e2e/java/teammates/e2e/cases/e2e/InstructorFeedbackSessionsPageE2ETest.java [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+        feedbackSessionsPage.verifyStatusMessage("The feedback session has been unpublished.");
+        feedbackSessionsPage.verifySessionDetails(openSession);
+        verifySessionPublishedState(openSession, false);
+        verifyEmailSent(studentToEmail.getEmail(), "TEAMMATES: Feedback session results unpublished"
+                + " [Course: " + copiedCourse.getName() + "][Feedback Session: "
+                + openSession.getFeedbackSessionName() + "]");
+
+        ______TS("download results");
+        feedbackSessionsPage.downloadResults(openSession);
+        ThreadHelper.waitFor(5000);
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/InstructorFeedbackSessionsPageE2ETest.java:179]: Maybe use retries here, where the success condition is the existence of downloaded file?

File: src/e2e/java/teammates/e2e/util/BackDoor.java [hunk 8]
```diff
@@ -520,10 +520,10 @@ public static FeedbackSessionAttributes getSoftDeletedSession(String feedbackSes
...
         FeedbackSessionsData sessionsData = JsonUtils.fromJson(response.responseBody, FeedbackSessionsData.class);
         FeedbackSessionData feedbackSession = sessionsData.getFeedbackSessions()
-                        .stream()
-                        .filter(fs -> fs.getFeedbackSessionName().equals(feedbackSessionName))
-                        .findFirst()
-                        .orElse(null);
+                .stream()
+                .filter(fs -> fs.getFeedbackSessionName().equals(feedbackSessionName))
+                .findFirst()
+                .orElse(null);
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/util/BackDoor.java:526]: Unrelated change?

File: src/main/java/teammates/ui/webapi/action/GetFeedbackSessionAction.java [hunk 9]
```diff
@@ -50,8 +52,13 @@ public ActionResult execute() {
         String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
         String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
         FeedbackSessionAttributes feedbackSession = logic.getFeedbackSession(feedbackSessionName, courseId);
-        Intent intent = Intent.valueOf(getNonNullRequestParamValue(Const.ParamsNames.INTENT));
 
+        if (feedbackSession == null) {
+            return new JsonResult("No feedback session with name " + feedbackSessionName
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/GetFeedbackSessionAction.java:57]: 👍

File: src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java [hunk 10]
```diff
@@ -146,4 +153,49 @@ protected AdminHomePage loginAdmin() {
...
+        FileHelper.deleteFile(filePath);
+    }
+
+    /**
+     * Verifies downloaded file has correct fileName and contains expected content.
+     */
+    protected void verifyDownloadedFile(String expectedContent, String expectedFileName) {
+        try {
+            String filePath = getTestDownloadsFolder() + expectedFileName;
+            assertTrue(FileHelper.readFile(filePath).contains(expectedContent));
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java:170]: Also, just checking one line may not be enough. Let's make `expectedContent` a list.

File: src/e2e/java/teammates/e2e/cases/e2e/InstructorFeedbackSessionsPageE2ETest.java (hunk 7 above)
Comment (jtankw3) [reply src/e2e/java/teammates/e2e/cases/e2e/InstructorFeedbackSessionsPageE2ETest.java]: Added it to the `verifyDownloadedFile` method

File: src/e2e/java/teammates/e2e/util/BackDoor.java (hunk 8 above)
Comment (jtankw3) [reply src/e2e/java/teammates/e2e/util/BackDoor.java:526]: I think it is because I realized it is the only method with a different indentation level. Should I not include it?

Comment (wkurniawan07) [reply src/e2e/java/teammates/e2e/util/BackDoor.java:526]: This should normally count as unrelated change, but I can close one eye.

**Review by wkurniawan07 — APPROVED** [review]
LGTM
